be/src/exec/partition/partition_hash_map.h

```diff
@@ -114,32 +144,48 @@ struct PartitionHashMapBase {
     // Each key mapped to a PartitionChunks which holds an array of chunks
     // chunk will be divided into multiply parts by partition key, and each parts will be append to the
     // last chunk of PartitionChunks.chunks. New chunk will be allocated if the last chunk reaches its capacity
-    // @key_loader used to load key for number type or slice type
-    template <typename HashMap, typename KeyLoader, typename KeyAllocator>
+    // Params:
+    // @key_loader
+    //      used to load key for number type or slice type.
+    // @new_partition_cb: void(size_t partition_idx)
+    //      called when coming a new key not in the hash map.
+    // @partition_chunk_consumer: void(size_t partition_idx, const ChunkPtr& chunk)
+    //      called for each partition with enough num rows after adding chunk to the hash map.
+    template <typename HashMap, typename KeyLoader, typename KeyAllocator, typename NewPartitionCallback,
+              typename PartitionChunkConsumer>
     void append_chunk_for_one_key(HashMap& hash_map, ChunkPtr chunk, KeyLoader&& key_loader,
-                                  KeyAllocator&& key_allocator, ObjectPool* obj_pool) {
+                                  KeyAllocator&& key_allocator, ObjectPool* obj_pool,
+                                  NewPartitionCallback&& new_partition_cb,
+                                  PartitionChunkConsumer&& partition_chunk_consumer) {
         if (is_downgrade) {
             return;
         }
+
         phmap::flat_hash_set<typename HashMap::key_type, typename HashMap::hasher, typename HashMap::key_equal,
                              typename HashMap::allocator_type>
                 visited_keys(chunk->num_rows());
         const auto size = chunk->num_rows();
+        auto next_partition_idx = hash_map.size();
         uint32_t i = 0;
+
         for (; !is_downgrade && i < size; i++) {
             const auto& key = key_loader(i);
             visited_keys.insert(key);
+
             bool is_new_partition = false;
             auto iter = hash_map.lazy_emplace(key, [&](const auto& ctor) {
                 is_new_partition = true;
-                return ctor(key_allocator(key), obj_pool->add(new PartitionChunks()));
+                auto* part_chunks = obj_pool->add(new PartitionChunks(next_partition_idx));
+                return ctor(key_allocator(key), part_chunks);
             });
             if (is_new_partition) {
                 check_downgrade(hash_map);
+                new_partition_cb(next_partition_idx++);
             }
+
             auto& value = *(iter->second);
             if (value.chunks.empty() || value.remain_size <= 0) {
-                if (!value.chunks.empty()) {
```

Comment:
what's the new condition means?

Reply:
If `value.select_indexes` is empty, then we also needn't call `append_selective`.